### ISSUE TITLE
World is not fully rendered in higher resolution

### ISSUE THREAD

**Issue by PatrikSamuelTauchim** [ZaneDubya/UltimaXNA#216]
If you set higher resolution in config, world is not rendered to fill the window.
![vystrizek2](https://cloud.githubusercontent.com/assets/11310587/8771983/29c31806-2ed0-11e5-9d4a-582215d586c4.JPG)

In dev\Ultima\World\WorldViews\IsometricView.cs is this code:

```
int renderDimensionY = 16; // the number of tiles that are drawn for half the screen (doubled to fill the entire screen).
int renderDimensionX = 18; // the number of tiles that are drawn in the x-direction ( + renderExtraColumnsAtSides * 2 ).
```

I can guess, it should be something like this.

```
int renderDimensionY = Settings.World.GumpResolution.Width / TILE_SIZE_INTEGER;
int renderDimensionX = Settings.World.GumpResolution.Height / TILE_SIZE_INTEGER;
```

After using this code, it looks like this. I remember that I already had this working before (long time ago) and had world rendered correctly in 1920x1080. Not sure what I changed, probably someting on how offset is calculated.
![vystrizek](https://cloud.githubusercontent.com/assets/11310587/8771934/ab804f86-2ecf-11e5-8317-1a73315ef0e2.JPG)

Can someone look at this? Thanks!

**Comment by ZaneDubya** [comment]
@Eskymak Definitely!

There is another limitation beyond the number of tiles drawn: The XNA Client only keeps 9x9 map blocks in memory (each block is 8x8 tiles). So there is definitely an upper limit to how much can be drawn: even at max resolution, only a diamond-shaped area approximately 72 tiles or ~3000 pixels across will be viewable.

**Comment by PatrikSamuelTauchim** [comment]
Thanks @ZaneDubya, 72 tiles (3K pixels) is more than enough for FullHD. There is also server side limitation on how far you can see (items, mobiles). On RunUO this could be probably easily changed to allow as see much more far.

For now, FullHD as maximum resolution should be enough. :-)

Can you point out what to change in actual code to fix this issue or you going to fix it yourself?

**Comment by ZaneDubya** [comment]
@Eskymak But that's a diamond-shaped area 3k pixels wide and 3k pixels high. Not a square. So there might be areas in the corners which wouldn't be covered by FullHD Resolution. Not sure. We might want to look into that. :)

You've already found the code that determines how large the display area is. The code to determine how that drawn area is offset is just a bit further down in the IsometricView.cs. All the code you should have to change is in lines 146-171.

**Comment by PatrikSamuelTauchim** [comment]
@ZaneDubya Yeah, I found that code already, but have no clue how it works and if it works correctly at all. Last time I had it working I spend like two days of randomly changing numbers in code and checking what changed in view... :-D

I set resolution to 1600x900 and result of that formula is renderOffset.X is 3176 and renderOffset.Y is -40800

What is the logic behind it? (i know, the formula) :-O

**Comment by ZaneDubya** [comment]
@Eskymak I see what you're saying. It's not exactly self-documenting code, and to be honest, it's been so long since I wrote the algorithm that its a little like a black box: it works, I don't touch it. :)

Here is what I know without getting too far into it:
1. IsometricView draws the world from the top right in rows from right to left. Rows proceed from top of the screen to bottom.
2. To get the location of any given tile on screen, the renderer takes these things into account: (i) where the tile would be if the entire world was drawn with tile (0,0) drawn at pixel (0,0); (ii) the amount these tiles should be offset given the location of the player; (iii) an additional offset to place the player at the center of the screen.
3. I believe that, at the time it is first set, renderOffset is the result of the calculations of all the factors from 2, supra.

**Comment by PatrikSamuelTauchim** [comment]
@ZaneDubya Thanks for info, still fighting.

Now I spotted this in my code...
int renderDimension**Y** = Settings.World.GumpResolution.**Width** / TILE_SIZE_INTEGER;
int renderDimension**X** = Settings.World.GumpResolution.**Height** / TILE_SIZE_INTEGER;

I guess that Width should not be Y... Yeah, you swapped order of X Y lines in your blackbox and I havent noticed that...

Now rendering, even in FullHD, is much better, just right part of screen is missing now. Thats progress... :-D

**Comment by naggie** [comment]
Good to see this! Being able to see more around me was one of the main reasons I used Kirrios' client (known as PlayUO at the time). That, and smooth motion -- not jumping per tile.

Not seeing mobiles past a certain distance isn't such an issue, if there was a dotted line or something to show the limitation it would be good.

**Comment by ZaneDubya** [comment]
@Eskymak How is work on the larger resolution coming? :)

**Comment by PatrikSamuelTauchim** [comment]
@ZaneDubya Now I was working on resolution setting in OptionsGump, so no progress on this. With renderDimension adjustment (above, but with correct X and Y) and setting resolutions to 1920x1080, it looks like this.

Not completely wrong, right side is missing,(probably offset issue) but you have to notice that repetitive lines of world not rendered (just sometimes and for unknown reason). I can guess that it has someting to do with that blocks? 
![ultimaxnafhd01](https://cloud.githubusercontent.com/assets/11310587/8835369/ae19442e-30c2-11e5-92a1-b3bcc70a62a7.jpg)

I will also need some help with that resolution setting in options gump. Otherwise it will be really ugly code. I had spent just two days on that dropdownlist... Check that branch. ;)

**Comment by ZaneDubya** [comment]
Ok! I've checked out your resolution code. I'll see if I can resolve any remaining bugs. We're definitely rendering more of the world than the server is sending to us at full hd resolution - I can see monsters, doors, etc. draw in as I walk! :)

Not sure what's causing the weird hitches in the landscape. Can you figure out what you do before they occur? Does it have something to do with walking one direction and then in the reverse direction perhaps?

**Comment by PatrikSamuelTauchim** [comment]
@ZaneDubya I will try to play with RunUO to check, if we can actually send more to the client in some easy way. I think there is just variable for that somewhere.

Regarding landscape, it probably really has something related to direction you are walking. On the screenshot I was walking from N to S. When walking in opposite direction, it seems fine. But needs more testing.

**Comment by PatrikSamuelTauchim** [comment]
@ZaneDubya Just note, the there is hardcoded 1920x1080 resolution in dev/Configuration/WorldSettings.cs for purpose of the testing, since this is only way how to change it, due to #210 

```
WindowResolution = new Resolution(1920, 1080);
GumpResolution = new Resolution(1920, 1080);
```

Also check my OptionsGump branch and take look at dropdownlist for resolution change, help is welcome, since I discovered an array just at that day and really dont feel confident enough to play with it. :-)

I guess that I can do following:
*\* get Index from dropdownlist
*\* get string from that index
*\* convert string to ints
*\* covert ints to Resolution

Doesnt look really nice to me and I believe that someone can do it in fancy way

**Comment by ZaneDubya** [comment]
@Eskymak I'm going to build on your changes to the resolution mode, probably have something up Fri or Sat. Thanks for working on this! With regards to your drop down code, perhaps @jeffboulanger can weigh in.

Here's one suggestion: you could build a List of Tuple<int, int> as you are filling the drop down list; each entry in the drop down list would have a corresponding Tuple in your resolution list. Here's some pseudocode:

```
Dropdown.Add("800x600");
ResolutionList.Add(new Tuple<int, int>(800, 600));
```

And when you select an entry:

```
Dropdown_OnChange()
{
    UltimaXNA.Resolution = ResolutionList[Dropdown.SelectedIndex].X, Y;
}
```

**Comment by PatrikSamuelTauchim** [comment]
@ZaneDubya Thanks for advise! I merged it with my ideas and it works. Resolution can now be set from OptionsGump and its correctly saved. Even the code looks nice to me. :-D

btw, I just made two arrays, one with strings for dropdown and second with actual Resolutions..

Seems like someting in rendering code is not updated correctly. When you save resolution change, view gets zoomed/unzoomed instead if resizing.

Check 2012e167b8bb1f1b6342c06d9c312a1768c66a40 and feel free to merge :-)

**Comment by ZaneDubya** [comment]
I'm continuing work on the better viewport fixing. I think I have it taken care of.

**Comment by ZaneDubya** [comment]
The maximum dimension (width or height) of the game window is 2048. This is not a limit we are going to be able to exceed while targeting the current XNA profile (reach). FullHD is _right_ on that limit. :)

**Comment by ZaneDubya** [comment]
<img width="1280" alt="sshd2" src="https://cloud.githubusercontent.com/assets/7041719/8884856/d575a414-321f-11e5-9492-79feb45d54a3.png">
And I think I've figured out why the tiles aren't loading correctly when you travel south...

**Comment by ZaneDubya** [comment]
Here's what I'm going to do: I'm going to limit the resolution of the game window to 2048x2048. And I'll increase the number of loaded map blocks (8x8 tiles) to enough to cover that entire window. Problem solved.

**Comment by ZaneDubya** [comment]
Done. And fixed those graphical errors you were seeing, @Eskymak.

**Comment by ZaneDubya** [comment]
Fixed in #225.
